### PR TITLE
feature/setup-gcloud-service-accounts

### LINKED ISSUE #352: Roles for CloudBuild service account to use SecretManager & Artifact Registry

### PR DESCRIPTION
**Pull request by abcarcol**
Closes #352

### CODE REVIEW

File: scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh [hunk 1]
```diff
@@ -83,18 +83,19 @@ else
...
+#Check if the service account exists
 if [ -n "$service_account" ];
 then
-    service_account_email="$service_account@$project_id.iam.gserviceaccount.com"
-    echo -e "${white}Checking if service account $service_account_email already exists..."
-    if ! gcloud iam service-accounts describe "$service_account_email" &> /dev/null;
+    echo -e "${white}Checking if service account $service_account already exists..."	
+    service_accounts=$(gcloud projects get-iam-policy "$project_id" --format='flattened' --format='flattened' | grep members | grep serviceAccount: | cut -d ':' -f 3-)
+    service_accounts_array=($service_accounts)
+    if [[ ! " ${service_accounts_array[*]} " =~ " ${service_account} " ]];
```
[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:92]: @sonatype-lift ignore

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 2]
```diff
@@ -95,35 +95,35 @@ fi
...
 if [ -n "$service_account" ];
-then
-    service_account_email="$service_account@$project_id.iam.gserviceaccount.com"
-    echo -e "${white}Checking if service account $service_account_email already exists..."
-    if ! gcloud iam service-accounts describe "$service_account_email" &> /dev/null;
+then	
+    echo -e "${white}Checking if service account $service_account already exists..."
+    service_accounts=$(gcloud projects get-iam-policy "$project_id" --format='flattened' | grep members | grep serviceAccount: | cut -d ':' -f 3-)
+    service_accounts_array=($service_accounts)
+    if [[ ! " ${service_accounts_array[*]} " =~ " ${service_account} " ]];
```
[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/setup-principal-account.sh:105]: @sonatype-lift ignore

File: scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh [hunk 3]
```diff
@@ -83,18 +83,19 @@ else
...
-#Check if the service account exists and if not, create it
+#Check if the service account exists
 if [ -n "$service_account" ];
 then
-    service_account_email="$service_account@$project_id.iam.gserviceaccount.com"
-    echo -e "${white}Checking if service account $service_account_email already exists..."
-    if ! gcloud iam service-accounts describe "$service_account_email" &> /dev/null;
+    echo -e "${white}Checking if service account $service_account already exists..."	
+    service_accounts=$(gcloud projects get-iam-policy "$project_id" --format='flattened' --format='flattened' | grep members | grep serviceAccount: | cut -d ':' -f 3-)
+    service_accounts_array=($service_accounts)
```
[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:91]: @sonatype-lift ignore

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 4]
```diff
@@ -95,35 +95,35 @@ fi
...
+
 if [ -n "$service_account" ];
-then
-    service_account_email="$service_account@$project_id.iam.gserviceaccount.com"
-    echo -e "${white}Checking if service account $service_account_email already exists..."
-    if ! gcloud iam service-accounts describe "$service_account_email" &> /dev/null;
+then	
+    echo -e "${white}Checking if service account $service_account already exists..."
+    service_accounts=$(gcloud projects get-iam-policy "$project_id" --format='flattened' | grep members | grep serviceAccount: | cut -d ':' -f 3-)
+    service_accounts_array=($service_accounts)
```
[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/setup-principal-account.sh:104]: @sonatype-lift ignore

File: scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:92]:
```suggestion
    if [[ ! " ${service_accounts_array[*]} " =~ " ${service_account} " ]]; # Searches right literal in left array. More info: https://stackoverflow.com/a/15394738
```

File: scripts/accounts/gcloud/setup-principal-account.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/setup-principal-account.sh:105]:
```suggestion
    if [[ ! " ${service_accounts_array[*]} " =~ " ${service_account} " ]]; # Searches right literal in left array. More info: https://stackoverflow.com/a/15394738
```

File: scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh [hunk 5]
```diff
@@ -83,18 +83,19 @@ else
...
+    if [[ ! " ${service_accounts_array[*]} " =~ " ${service_account} " ]]; # Searches right literal in left array. More info: https://stackoverflow.com/a/15394738
```
[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:92]: @sonatype-lift ignore

**Review by albertdb — CHANGES_REQUESTED** [review]
I see this as a fix to overcome a limitation that worsens user experience, that now has to know the format of a service account email to create a new one.

We can ask for Service Account name **or* email. And if the value does not contain @ we assume it is a custom service account as before.

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 6]
```diff
@@ -23,14 +23,14 @@ while getopts g:s:p:r:f:c:i:k:h flag
...
         r) roles=${OPTARG};;
         f) roles_file=${OPTARG};;
         c) custom_role_file=${OPTARG};;
-        k) path_store_key=${OPTARG};;
-	i) custom_role_id=${OPTARG};;
-	h) helpFunction ;;
+	    i) custom_role_id=${OPTARG};;
+	    k) key_path=${OPTARG};;
+	    h) helpFunction ;;
         ? ) helpFunction ;; # Print helpFunction in case parameter is non-existent.
```
Comment (albertdb) [inline scripts/accounts/gcloud/setup-principal-account.sh:34]: Please replace here and across all script tabs by spaces, keeping indentation consistent.

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 7]
```diff
@@ -94,39 +94,50 @@ else
...
-    if ! gcloud iam service-accounts keys create $path_store_key --iam-account="$service_account_email" &> /dev/null;
-    then      
-        echo -e "${red}Error: Service account key could not be created." >&2
-        echo -ne "${white}"
-	exit 2
-    else
-        echo -e "${green}Service account key creation ended successfully."
-	echo -ne "${white}"
+        echo -e "${white}Checking if service account $service_account already exists..."
+        if [[ ! " ${service_accounts_array[*]} " =~ " ${service_account} " ]];
```
[the remark this answers is not in the document]
Comment (albertdb) [reply scripts/accounts/gcloud/setup-principal-account.sh:106]: Does this work when is the first or last element of the array?

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 8]
```diff
@@ -94,39 +94,50 @@ else
...
+        then
+            echo -e "${red}Error: Service account $service_account. does not exist" >&2
+	        echo -ne "${white}"
+	        exit 2
+        fi
+    else #User-managed service account
+        service_account_email="$service_account@$project_id.iam.gserviceaccount.com"
+        echo -e "${white}Checking if service account $service_account already exists..."
+        if [[ ! " ${service_accounts_array[*]} " =~ " $service_account_email " ]];
```
[the remark this answers is not in the document]
Comment (albertdb) [reply scripts/accounts/gcloud/setup-principal-account.sh:115]: Does this work when is the first or last element of the array?

File: scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh [hunk 9]
```diff
@@ -9,7 +9,7 @@ helpFunction()
     echo ""
     echo "Arguments:"
     echo "  -g     [Required] Google Account of an end user. Mutually exclusive with -s."
-    echo "  -s     [Required] Service Account name. Mutually exclusive with -g."
+    echo "  -s     [Required] Service Account email. Mutually exclusive with -g."
```
Comment (albertdb) [inline scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:12]: Please accept both:

````
[[ "$serviceAccount" =~ .*"@".* ]] && service_account_email="$service_account@$project_id.iam.gserviceaccount.com" || service_account_email="$serviceAccount"
````

File: scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh [hunk 10]
```diff
@@ -83,18 +83,19 @@ else
...
 fi
 
-#Check if the service account exists and if not, create it
+#Check if the service account exists
 if [ -n "$service_account" ];
 then
-    service_account_email="$service_account@$project_id.iam.gserviceaccount.com"
-    echo -e "${white}Checking if service account $service_account_email already exists..."
-    if ! gcloud iam service-accounts describe "$service_account_email" &> /dev/null;
+    echo -e "${white}Checking if service account $service_account already exists..."
```
Comment (albertdb) [inline scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:89]: Does `gcloud iam service-accounts describe` not work for google managed service accounts?

**Review by albertdb — CHANGES_REQUESTED** [review]

File: scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh (hunk 10 above)
Comment (abcarcol) [reply scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:89]: Oops I totally forgot to fix this script. Sorry!

File: scripts/accounts/gcloud/setup-principal-account.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/setup-principal-account.sh:106]: Yes, it works but I do not understand why do you ask it. Please, elaborate a little on this.

[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/setup-principal-account.sh:106]: @sonatype-lift ignore

File: scripts/accounts/gcloud/setup-principal-account.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/setup-principal-account.sh:115]: Same as previous comment

[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/setup-principal-account.sh:115]: @sonatype-lift ignore

File: scripts/accounts/gcloud/setup-principal-account.sh (hunk 6 above)
Comment (abcarcol) [reply scripts/accounts/gcloud/setup-principal-account.sh:34]: Done

File: scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh (hunk 9 above)
Comment (abcarcol) [reply scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:12]: I think you meant this:
```suggestion
    [[ "$serviceAccount" =~ .*"@".* ]] && service_account_email="$serviceAccount" || service_account_email="$service_account@$project_id.iam.gserviceaccount.com"  
```

Comment (abcarcol) [reply scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:12]: I forgot about this script, now it is done.

File: scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh (hunk 10 above)
Comment (abcarcol) [reply scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:89]: Yes, it works for any kind of service account, the code was not changed. Now it is done.

**Review by lpinon — DISMISSED** [review]

File: scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh [hunk 11]
```diff
@@ -83,12 +83,14 @@ else
...
 if [ -n "$service_account" ];
 then
-    service_account_email="$service_account@$project_id.iam.gserviceaccount.com"
-    echo -e "${white}Checking if service account $service_account_email already exists..."
-    if ! gcloud iam service-accounts describe "$service_account_email" &> /dev/null;
+    [[ "$service_account" =~ .*"@".* ]] && service_account_email="$service_account" || service_account_email="$service_account@$project_id.iam.gserviceaccount.com"
+    echo -e "${white}Checking if service account $service_account_email already exists..."	
+    service_accounts=$(gcloud projects get-iam-policy "$project_id" --format='flattened' --format='flattened' | grep members | grep serviceAccount: | cut -d ':' -f 3-)
+    service_accounts_array=($service_accounts)
+    if [[ ! " ${service_accounts_array[*]} " =~ " ${service_account_email} " ]]; # Searches right literal in left array. More info: https://stackoverflow.com/a/15394738
```
[the remark this answers is not in the document]
Comment (abcarcol) [reply scripts/accounts/gcloud/verify-principal-roles-and-permissions.sh:93]: @sonatype-lift ignore

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 12]
```diff
@@ -86,47 +86,58 @@ else
     then
         echo -e "${red}Error: Could not set current project to $project_id." >&2
         echo -ne "${white}"
-	exit 2
+    exit 2
```
Comment (albertdb) [inline scripts/accounts/gcloud/setup-principal-account.sh:89]:
```suggestion
        exit 2
```

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 13]
```diff
@@ -147,14 +163,14 @@ then
     IFS=',' read -ra roles_array <<< "$roles"
     for role_to_check in "${roles_array[@]}"; do
         if ! gcloud projects add-iam-policy-binding "$project_id" --member="$memberValue" --role="$role_to_check" &> /dev/null;
-	then
-	    echo -e "${red}Error: Attaching role $role_to_check to $memberValue for project $project_id." >&2
-	    echo -ne "${white}"
+    then
+        echo -e "${red}Error: Attaching role $role_to_check to $memberValue for project $project_id." >&2
+        echo -ne "${white}"
```
Comment (albertdb) [inline scripts/accounts/gcloud/setup-principal-account.sh:168]:
```suggestion
        then
            echo -e "${red}Error: Attaching role $role_to_check to $memberValue for project $project_id." >&2
            echo -ne "${white}"
```

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 14]
```diff
@@ -147,14 +163,14 @@ then
...
             exit 2
-	else
-	    echo -e "${green}Attached role $role_to_check to $memberValue in project $project_id."
-	    echo -ne "${white}"
-	fi
+    else
+        echo -e "${green}Attached role $role_to_check to $memberValue in project $project_id."
+        echo -ne "${white}"
+    fi
```
Comment (albertdb) [inline scripts/accounts/gcloud/setup-principal-account.sh:173]:
```suggestion
        else
            echo -e "${green}Attached role $role_to_check to $memberValue in project $project_id."
            echo -ne "${white}"
        fi
```

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 15]
```diff
@@ -166,11 +182,11 @@ then
...
-	then
-	    echo -e "${red}Error: Attaching role $role_to_check to $memberValue in project $project_id." >&2
-	    echo -ne "${white}"
-	    exit 2
-	else
+    then
+        echo -e "${red}Error: Attaching role $role_to_check to $memberValue in project $project_id." >&2
+        echo -ne "${white}"
+        exit 2
+    else
```
Comment (albertdb) [inline scripts/accounts/gcloud/setup-principal-account.sh:189]:
```suggestion
        then
            echo -e "${red}Error: Attaching role $role_to_check to $memberValue in project $project_id." >&2
            echo -ne "${white}"
            exit 2
        else
```

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 16]
```diff
@@ -179,23 +195,23 @@ fi
...
 if [ -n "$custom_role_file" ];
-then	
+then
     echo -e "${white}Creating role $custom_role_id defined in $custom_role_file for project $project_id..."
     if ! gcloud iam roles create "$custom_role_id" --project="$project_id" --file="$custom_role_file" &> /dev/null;
     then
         echo -e "${red}Error: Creating custom role $custom_role_id for project $project_id." >&2
         echo -ne "${white}"
-	exit 2
+    exit 2
```
Comment (albertdb) [inline scripts/accounts/gcloud/setup-principal-account.sh:204]:
```suggestion
        exit 2
```

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 17]
```diff
@@ -179,23 +195,23 @@ fi
...
     else
-	echo -e "${green}Created custom role $custom_role_id for project $project_id successfully."
-	echo -ne "${white}"
+    echo -e "${green}Created custom role $custom_role_id for project $project_id successfully."
+    echo -ne "${white}"
```
Comment (albertdb) [inline scripts/accounts/gcloud/setup-principal-account.sh:207]:
```suggestion
        echo -e "${green}Created custom role $custom_role_id for project $project_id successfully."
        echo -ne "${white}"
```

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 18]
```diff
@@ -179,23 +195,23 @@ fi
...
     fi
     echo -e "${white}Binding role $custom_role_id to principal $memberValue in project $project_id..."
     if ! gcloud projects add-iam-policy-binding "$project_id" --member="$memberValue" --role=projects/"$project_id"/roles/"$custom_role_id" &> /dev/null;
     then
         echo -e "${red}Error: Attaching custom role $custom_role_id to $memberValue in project $project_id." >&2
         echo -ne "${white}"
-	exit 2
+    exit 2
```
Comment (albertdb) [inline scripts/accounts/gcloud/setup-principal-account.sh:214]:
```suggestion
        exit 2
```

**Review by albertdb — CHANGES_REQUESTED** [review]
Fixed with suggestions one script. Please fix the other, it has the same issue.

**Review by albertdb — APPROVED** [review]

**Pull request merged**